### DISCUSSION TITLE
Concatenate multiple FMUs with NN

### DISCUSSION THREAD

**Discussion by awecefil** [ThummeTo/FMIFlux.jl#140] [Q&A]
**Problem Description:**

I want to know if FMIFlux can currently chain multiple FMUs and NNs together. For example, I have constructed a circuit system in OpenModelica that includes a voltage source and a resistor. Using OpenModelica, I split this system into three subsystems (voltage source, resistor, and ground), and exported each as an independent FMU. I then tried importing these three subsystem FMUs into OpenModelica and connecting their inputs/outputs for simulation. Unsurprisingly, the simulation results were the same as the original circuit system. 
Diagram as follow, from left to right, it sequentially shows the Voltage Source FMU, Resistor FMU, and Ground FMU:
![image](https://github.com/ThummeTo/FMIFlux.jl/assets/16224322/5aeece9a-82ef-4f4d-8b55-9a53eb4ead17)

**Now, I want to know two things:**

1. Is it possible to achieve the same functionality described above in FMIFlux or other FMI packages?
2. Is it possible in FMIFlux to chain NNs and multiple FMUs together? For example, I want to replace the original linear resistor with an NN and use training data to simulate a nonlinear resistor.

**Current Attempts:**

I found that FMIFlux currently supports using Parallel to chain multiple FMUs and NNs (via Chain). However, after looking into it in detail, I discovered that the Parallel function allows multiple FMUs to run concurrently and produce results. We can then concatenate the results or perform other operations like summation. However, this is different from sequentially chaining multiple subsystem FMUs as I intend. 

If there is any misunderstanding on my part, please correct me. Thank you.

**Comment by ThummeTo** [comment]
Exactly, there are two to different operation modes: 
- "co-simulation" which uses different solvers to solve the models independently of each other and syncs signals after every time step
- "model integration" which actually combines the two ODE systems (or ME-FMUs) into a single one and then solves the combined system.

As you correctly noticed, the parallel example simulates the FMUs in CS-mode.
From the text, it sounds like you want to do "model-integration", but importing FMUs in OpenModelica and simulate them together sounds like co-simulation to me....

Can you specify:
(a) if you want to operate the FMUs in CS or ME mode?
(b) if ME: if you want to use the same solver for both FMUs (model integration) or different solvers for every FMU?
(c) do you want to connect states/state derivatives or inputs/outputs?

In any case, I think there is a solution for your problem, but it is easier to discuss if we further specify.

Best regards!

PS: I convert this to discussion :-)

**Reply by awecefil** [reply]
Hi @ThummeTo, thanks for your reply! Here is the specification about my problem：

(a) if you want to operate the FMUs in CS or ME mode?
     → I prefer operate in ME mode because I have no demand on using different solvers for different FMU.
(b) if ME: if you want to use the same solver for both FMUs (model integration) or different solvers for every FMU?
     → Currently, I want to use the same solver for every FMU.
(c) do you want to connect states/state derivatives or inputs/outputs?
     →  I think in my case is  inputs/outputs but sorry I am not sure about this, could you further explain the difference between them?

What I actually want to do is **embed a neural network within a system** to replace certain electronic components (usually components whose physical behavior is difficult to model, such as nonlinear resistors). 
Based on this direction, if I want to embed a neural network into the simulation model, I first want to separate each circuit component except the resistor and convert them into FMUs (each FMU being a subsystem or component). Then, I hope to use FMIFlux.jl to chain these FMUs and the neural network together to form a complete system for subsequent training. However, I am currently unsure if FMIFlux can chain "multiple FMUs" and NN together into a complete system.

By the way, if I have other problems about FMIFlux, should I open a new discussion or issue? It is about：
1. CS-FMU with FMUParameterRegistrator (works fine when using ME-FMU)
2. FMIFlux.train! error when using FMU exported from OpenModelica (no error when using built-in FMU exported from Dymola)

**Reply by awecefil** [reply]
Hi, I found a way to avoid concatenating multiple FMUs with NN like I said above to achieve my demand, so this problem has almost be solved.

However, the another two problems I mentioned in my last reply still remain, should I open a new discussion or issue?

**Reply by ThummeTo** [reply]
1. Looks like an **issue** to me (so feel free to open an issue with MWE)
2. It is most likely a new **discussion**. Almost always the FMU generates an error, that leads to the visible problem. A goof check is if you can simulate the pure FMU with FMI.jl. Also enable logging for the FMU, to see warnings/errors.
